src/System.Memory/tests/Memory/ToEnumerable.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Xunit;
+using System.Linq;
+
+namespace System.MemoryTests
+{
+    public static partial class MemoryTests
+    {
+        [Fact]
+        public static void ToEnumerable()
+        {
+            int[] a = { 91, 92, 93 };
+            var memory = new Memory<int>(a);
+            IEnumerable<int> copy = memory.ToEnumerable();
+            Assert.Equal<int>(a, copy);
+        }
+
+        [Fact]
+        public static void ToEnumerableWithIndex()
+        {
+            int[] a = { 91, 92, 93, 94, 95 };
+            var memory = new Memory<int>(a);
+            IEnumerable<int> copy = memory.Slice(2).ToEnumerable();
+
+            Assert.Equal<int>(new int[] { 93, 94, 95 }, copy);
+        }
+
+        [Fact]
+        public static void ToEnumerableWithIndexAndLength()
+        {
+            int[] a = { 91, 92, 93 };
+            var memory = new Memory<int>(a, 1, 1);
+            IEnumerable<int> copy = memory.ToEnumerable();
+            Assert.Equal<int>(new int[] { 92 }, copy);
+        }
+
+        [Fact]
+        public static void ToEnumerableEmpty()
+        {
+            Memory<int> memory = Memory<int>.Empty;
+            IEnumerable<int> copy = memory.ToEnumerable();
+            Assert.Equal(0, copy.Count());
+        }
+
+        [Fact]
+        public static void ToEnumerableDefault()
+        {
+            Memory<int> memory = default;
+            IEnumerable<int> copy = memory.ToEnumerable();
+            Assert.Equal(0, copy.Count());
+        }
+
+        [Fact]
+        public static void ToEnumerableForEach()
+        {
+            int[] a = { 91, 92, 93 };
+            var memory = new Memory<int>(a);
+            int index = 0;
+            foreach (int curr in memory.ToEnumerable())
+            {
+                Assert.Equal(a[index++], curr);
+            }
+        }
+
+        [Fact]
+        public static void ToEnumerableGivenToExistingConstructor()
+        {
+            int[] a = { 91, 92, 93 };
+            var memory = new Memory<int>(a);
+            IEnumerable<int> enumer = memory.ToEnumerable();
+            var li = new List<int>(enumer);
+            Assert.Equal(a, li);
+        }
+    }
+}
```

Comment:
Consider adding a test that Assert.Sames on the enumerable and its enumerator.

Reply:
I added one, and it's failing using the iterator. 

I glanced at the IL using ILSpy to see how the iterator was being unfolded but ILSpy showed me the iterator C#, so that wasn't super helpful.

Looking at the Roslyn code, the only apparent case in which `GetEnumerator` returns ```this``` is [```if (this.initialThreadId == Thread.CurrentThread.ManagedThreadId && this.state == -2)```](https://github.com/dotnet/roslyn/blob/701ca5b2ce7e331d3bb47e55d842a8a1634e501d/src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs#L259). The state is initialized to `-2` and set to `0` after the first `GetEnumerator` call is made which also means that subsequent `GetEnumerator` calls will create new instances of the `MemoryEnumerable`. It should mean that the first `GetEnumerator` call returns `this`, but that doesn't appear to be happening. Possibly because of the threadID check I guess.

Reply:
Can you share the test code?

Reply:
```
        [Fact]
        public static void ToEnumerableSameAsIEnumerator()
        {
            int[] a = { 91, 92, 93 };
            var memory = new Memory<int>(a);
            IEnumerable<int> enumer = memory.ToEnumerable();
            IEnumerator<int> enumerat = enumer.GetEnumerator();
            Assert.Same(enumer, enumerat);
        }
```

Reply:
@jaredpar, @VSadov, has something broken here in Roslyn recently?  The optimization for reusing the enumerable as an enumerator appears to be gone; in both debug and release builds, I see the GetEnumerator method here doing:
```
.method private hidebysig newslot virtual final 
        instance class [System.Private.CoreLib]System.Collections.Generic.IEnumerator`1<!T> 
        'System.Collections.Generic.IEnumerable<T>.GetEnumerator'() cil managed
{
  .custom instance void [System.Private.CoreLib]System.Diagnostics.DebuggerHiddenAttribute::.ctor() = ( 01 00 00 00 ) 
  .override  method instance class [System.Private.CoreLib]System.Collections.Generic.IEnumerator`1<!0> class [System.Private.CoreLib]System.Collections.Generic.IEnumerable`1<!T>::GetEnumerator()
  // Code size       21 (0x15)
  .maxstack  2
  .locals init (class System.MemoryExtensions/'<ToEnumerable>d__46`1'<!T> V_0)
  IL_0000:  ldc.i4.0
  IL_0001:  newobj     instance void class System.MemoryExtensions/'<ToEnumerable>d__46`1'<!T>::.ctor(int32)
  IL_0006:  stloc.0
  IL_0007:  ldloc.0
  IL_0008:  ldarg.0
  IL_0009:  ldfld      valuetype [System.Private.CoreLib]System.Memory`1<!0> class System.MemoryExtensions/'<ToEnumerable>d__46`1'<!T>::'<>3__memory'
  IL_000e:  stfld      valuetype [System.Private.CoreLib]System.Memory`1<!0> class System.MemoryExtensions/'<ToEnumerable>d__46`1'<!T>::memory
  IL_0013:  ldloc.0
  IL_0014:  ret
} // end of method '<ToEnumerable>d__46`1'::'System.Collections.Generic.IEnumerable<T>.GetEnumerator'
```
This is with 2.6.0.62316 (fbe99784).

Reply:
Is `Environment.GetManagedThreadId` available in the project? 
ThreadId is needed for this optimization to work.

Reply:
I see.  @ianhays, what happens if you add:
```XML
<ProjectReference Include="..\..\System.Runtime.Extensions\src\System.Runtime.Extensions.csproj" />
```
to the:
```XML
<ItemGroup Condition="'$(IsPartialFacadeAssembly)' == 'true'">
```
section of the src System.Memory.csproj?

Reply:
The test still fails.

Reply:
I played around with it some more and can't get the iterator to compile the way I want it to.

I tried writing it in C# with the IEnumerable/IEnumerator as one class, but attempting to mimic the threading behavior raises issues that "Thread.CurrentThread" is "inaccessible due to its protection level". I'm guessing that's why the iterator was having the issues.

Something about it being an extension method or a partial facade assembly, maybe? I don't yet understand why `Thread.CurrentThread` is complaining as inaccessible.

Reply:
What about `Environment.CurrentManagedThreadID` - is that also inaccessible?

Reply:
> The test still fails.

@ianhays, really? Hmm.  I checked out the commit, made that change, and the test then passes for me.